### PR TITLE
Support deployment of a joined NooBaa on a remote cluster 

### PR DESCRIPTION
**Pull request by nb-ohad**
Currently, we mainly deploy an S3 service and an auto-scaled endpoint group

Changes:
- Create an auth_token secret to be used by endpoints
- Introduce a join secret to hold information on how to join an existing NooBaa cluster
- Add a join secret reference on the NooBaa CRD to indicate that the deployed NooBaa should join an existing cluster using the join information specified in the referenced secret

### CODE REVIEW

File: deploy/internal/deployment-endpoint.yaml [hunk 1]
```diff
@@ -46,16 +46,22 @@ spec:
...
-        - name: JWT_SECRET
-          valueFrom:
-            secretKeyRef:
-              name: noobaa-server
-              key: jwt
         - name: MGMT_URL
+        - name: MGMT_PORT
+        - name: BG_PORT
+        - name: MD_PORT
+        - name: HOSTED_AGENTS_PORT
```
Comment (dannyzaken) [inline deploy/internal/deployment-endpoint.yaml]: do we need `HOSTED_AGENTS_PORT`?

File: pkg/apis/noobaa/v1alpha1/noobaa_types.go [hunk 2]
```diff
@@ -121,7 +121,13 @@ type NooBaaSpec struct {
 
 	// Endpoints (optional) sets configuration info for the noobaa endpoint
 	// deployment.
+	// +optional
 	Endpoints *EndpointsSpec `json:"endpoints,omitempty"`
+
+	// JoinSecret (optional) instracts the operator to join another cluster
```
Comment (dannyzaken) [inline pkg/apis/noobaa/v1alpha1/noobaa_types.go]: typo - ~~instracts~~ -> instructs

File: deploy/crds/noobaa.io_noobaas_crd.yaml [hunk 3]
```diff
@@ -753,6 +753,19 @@ spec:
                     TODO: Add other useful fields. apiVersion, kind, uid?'
                   type: string
               type: object
+            joinSecret:
+              description: JoinSecret (optional) instracts the operator to join another
```
Comment (nimrod-becker) [inline deploy/crds/noobaa.io_noobaas_crd.yaml]:
```suggestion
              description: JoinSecret (optional) instructs the operator to join another
```

File: pkg/system/phase2_creating.go [hunk 4]
```diff
@@ -31,41 +31,57 @@ func (r *Reconciler) ReconcilePhaseCreating() error {
 		"noobaa operator started phase 2/4 - \"Creating\"",
 	)
 
-	// A failure to discover OAuth endpoints should not fail the entire reconcile phase.
-	oAuthEndpoints, err := util.DiscoverOAuthEndpoints()
-	if err != nil {
-		r.Logger.Warnf("Discovery of OAuth endpoints failed, got: %v", err)
+	if r.JoinSecret == nil {
```
Comment (nimrod-becker) [inline pkg/system/phase2_creating.go]: Isn't it better to split into two ReconcilePhaseCreating? One for Join nil and for for !nil?

File: deploy/internal/deployment-endpoint.yaml (hunk 1 above)
Comment (nimrod-becker) [reply deploy/internal/deployment-endpoint.yaml]: +1 , Assuming we work with n2n_proxy we probably don't right?

File: pkg/system/phase4_configuring.go [hunk 5]
```diff
@@ -31,10 +31,16 @@ func (r *Reconciler) ReconcilePhaseConfiguring() error {
 		"noobaa operator started phase 4/4 - \"Configuring\"",
 	)
 
-	if err := r.ReconcileSecretOp(); err != nil {
-		return err
+	if r.JoinSecret == nil {
```
Comment (nimrod-becker) [inline pkg/system/phase4_configuring.go]: Same as the comment on phase2? Maybe also 3 (even though its not a lot of changes)

File: pkg/system/phase2_creating.go [hunk 6]
```diff
@@ -31,41 +31,57 @@ func (r *Reconciler) ReconcilePhaseCreating() error {
...
+		// A failure to discover OAuth endpoints should not fail the entire reconcile phase.
+		oAuthEndpoints, err := util.DiscoverOAuthEndpoints()
```
Comment (dannyzaken) [inline pkg/system/phase2_creating.go]: should this be inside the `if`?

File: pkg/system/phase2_creating.go (hunk 4 above)
Comment (guymguym) [reply pkg/system/phase2_creating.go]: sounds better to me too

Comment (nb-ohad) [reply pkg/system/phase2_creating.go]: @nimrod-becker @guymguym 
I don't think so. It creates duplicate code that needs to be updated for every change in both places.
Also, the end goal (in the future) is to have full NooBaa on the remote cluster which means 
that we will take reconcile function out of the if statements little by little

File: deploy/internal/deployment-endpoint.yaml (hunk 1 above)
Comment (nb-ohad) [reply deploy/internal/deployment-endpoint.yaml]: These ports are moved to the other cluster to support a full router configuration, even if currently we do not invoke any API calls that are targeted in the hosted agents.
In any case, we still need to provide the server_rpc router an address (which includes a port) for the hosted_agents domain on the router

File: pkg/system/phase4_configuring.go (hunk 5 above)
Comment (nb-ohad) [reply pkg/system/phase4_configuring.go]: Please see my answer above

File: pkg/system/phase2_creating.go (hunk 4 above)
Comment (dannyzaken) [reply pkg/system/phase2_creating.go]: at the moment it's not very readable. there is no need to duplicate code. if there is a lot of common code you can extract it to a function

Comment (nimrod-becker) [reply pkg/system/phase2_creating.go]: lets not create duplicate code. Create a common base which runs in both cases and then a specific func for each case.

File: deploy/internal/deployment-endpoint.yaml (hunk 1 above)
Comment (nb-ohad) [reply deploy/internal/deployment-endpoint.yaml]: Not using ports anymore - Changed to use full address for each service

File: pkg/system/phase2_creating.go (hunk 4 above)
Comment (nb-ohad) [reply pkg/system/phase2_creating.go]: I changed the way the code is structured. The main change is that each reconcile function checks for itself if it needs to be skipped

**Pull request merged**